File: services/confluence-importer/src/pipeline/chunk-batch.ts
import type { EmbeddingClient } from "../embeddings/index.js";
import type { PgVectorClient } from "../pgvector/index.js";
import type { Chunk } from "../chunking/types.js";
import type { Logger } from "../logger.js";

export async function processChunkBatch(
  chunks: Chunk[],
  embedding: EmbeddingClient,
  store: PgVectorClient,
  logger: Logger | undefined,
): Promise<void> {
  logger?.debug("Embedding chunk batch", { count: chunks.length });
  const texts = chunks.map((c) => c.content);
  const embeddings = await embedding.embed(texts);
  await store.upsertChunks(chunks, embeddings);
  logger?.debug("Stored chunk batch in vector store", { count: chunks.length });
}

File: services/confluence-importer/src/confluence_importer/pipeline/chunk_batch.py
"""Page-buffer flushing shared by the ingest and local-file pipelines.

Pages are buffered whole so embedding calls stay batched across pages while
each page's chunks are still swapped atomically (embed first, then
``replace_page_chunks`` in one transaction).
"""

import logging

from confluence_importer.chunking.base import Chunk
from confluence_importer.embeddings.base import EmbeddingProvider
from confluence_importer.store.pgvector import PgVectorStore

logger = logging.getLogger(__name__)

PageChunks = tuple[str, list[Chunk]]


def flush_page_buffer(
    buffer: list[PageChunks],
    embedding: EmbeddingProvider,
    store: PgVectorStore,
) -> None:
    """Embed all buffered chunks in one batch, then replace each page atomically."""
    if not buffer:
        return

    pages = list(buffer)
    buffer.clear()

    texts = [chunk.content for _, chunks in pages for chunk in chunks]
    logger.debug("Embedding page buffer (pages=%d, chunks=%d)", len(pages), len(texts))
    embeddings = embedding.embed_texts(texts) if texts else []

    offset = 0
    for page_id, chunks in pages:
        store.replace_page_chunks(page_id, chunks, embeddings[offset : offset + len(chunks)])
        offset += len(chunks)
